src/sync/index.js

```diff
@@ -20,6 +31,89 @@ export type SyncDatabaseChangeSet = $Exact<{ [TableName<any>]: SyncTableChangeSe
 
 export type SyncLocalChanges = $Exact<{ changes: SyncDatabaseChangeSet, affectedRecords: Model[] }>
 
+// *** Applying remote changes ***
+
+const getIds = map(({ id }) => id)
+const idsForChanges = ({ created, updated, deleted }: SyncTableChangeSet): RecordId[] => [
+  ...getIds(created),
+  ...getIds(updated),
+  ...deleted,
+]
+const queryForChanges = changes => Q.where(columnName('id'), Q.oneOf(idsForChanges(changes)))
+
+const findRecord = <T: Model>(id: RecordId, list: T[]) => find(record => record.id === id, list)
+
+function applyRemoteChangesToCollection<T: Model>(
+  collection: Collection<T>,
+  changes: SyncTableChangeSet,
+): Promise<void> {
+  const { database, table } = collection
+  return database.action(async () => {
+    const { created, updated, deleted: deletedIds } = changes
+
+    const records = await collection.query(queryForChanges(changes)).fetch()
+    const locallyDeletedIds = await database.adapter.getDeletedRecords(table)
+
+    // Destroy records (if already marked as deleted, just destroy permanently)
+    const recordsToDestroy = filter(record => contains(record.id, deletedIds), records)
+    const deletedRecordsToDestroy = filter(id => contains(id, deletedIds), locallyDeletedIds)
+
+    await allPromises(record => record.destroyPermanently(), recordsToDestroy)
+    await database.adapter.destroyDeletedRecords(collection.table, deletedRecordsToDestroy)
+
+    // Insert and update records
+    const recordsToInsert = map(raw => {
+      const currentRecord = findRecord(raw.id, records)
+      if (currentRecord) {
+        // TODO: log error -- record already exists, update instead
+        return prepareUpdateFromRaw(currentRecord, raw)
+      } else if (contains(raw.id, locallyDeletedIds)) {
+        // FIXME: this will fail
+        // database.adapter.destroyDeletedRecords(collection.table, raw.id)
+        return prepareCreateFromRaw(collection, raw)
+      }
+
+      return prepareCreateFromRaw(collection, raw)
+    }, created)
+
+    const recordsToUpdate = map(raw => {
+      const currentRecord = findRecord(raw.id, records)
+
+      if (currentRecord) {
+        return prepareUpdateFromRaw(currentRecord, raw)
+      } else if (contains(raw.id, locallyDeletedIds)) {
+        // Nothing to do, record was locally deleted, deletion will be pushed later
+        return null
+      }
+
+      // Record doesn't exist (but should) — just create it
+      return prepareCreateFromRaw(collection, raw)
+    }, updated).filter(Boolean)
```

Comment:
why do you mix up `rambdax` functions and native methods?

```
const recordsToUpdate = piped(
  updated,
  map(…),
  filter(Boolean)
)
```

does it look good? 😉